src/compiler/types.ts

```diff
@@ -8685,29 +8685,74 @@ export interface TransformationResult<T extends Node> {
  * A function that is used to initialize and return a `Transformer` callback, which in turn
  * will be used to transform one or more nodes.
  */
-export type TransformerFactory<T extends Node> = (context: TransformationContext) => Transformer<T>;
+export type TransformerFactory<TIn extends Node, TOut extends Node | undefined = TIn> = (context: TransformationContext) => Transformer<TIn, TOut>;
 
 /**
  * A function that transforms a node.
  */
-export type Transformer<T extends Node> = (node: T) => T;
+export type Transformer<TIn extends Node, TOut extends Node | undefined = TIn> = (node: TIn) => TOut;
 
 /**
  * A function that accepts and possibly transforms a node.
  */
-export type Visitor = (node: Node) => VisitResult<Node>;
+export type Visitor<TIn extends Node = Node, TOut extends Node | undefined = TIn | undefined> = (node: TIn) => VisitResult<TOut>;

```

Comment:
`Visitor` was always weak because we were supposed to depend on the `test` type guard to verify the output. Was this change just to make `visitNode`/`visitNodes` more strongly typed like a `map<T, U>(...)` callback?

Reply:
This change was to make `visitNode` and `visitNodes` accurately represent when they can return undefined. Without this, there's no way to statically determine that a visitor returns undefined or not, because the old type always has it there.